Handle missing data

Murphydbuffalo opened the pull request:
This PR is probably best to review this PR commit-by-commit and does a few things:
1. Gracefully handles attempts to make API calls with a missing participant ID or experiment ID. Now the client won't even attempt the API call if one of those parameters is missing. I wanted to do this not as a matter of optimization, but rather to make it easy to opt certain users out of experiments. Eg for our landing page experiment, if the user in question has already signed up, I don't want them to participate in the experiment at all, as that would throw off the numbers. With this change that sort of thing is easy to do: just have the function for getting the user's participant ID return `None` if the user exists in the database.
2. DRYs out the `try_conversion` and `try_assignment` methods which have been getting more and more complex and share  the vast majority of their functionality.
3. Adds more details to the logging that `try_assignment` and `try_conversion` do.
4. Explicitly sets the content type of the client's requests to `application/json`. Not having done this was causing the Hyp API to [return HTML error messages in some cases due to this code](https://github.com/Murphydbuffalo/hyp2/blob/main/web/hyp/views/shared_views.py#L6). Because I made this change I also had to re-record our HTTP requests in the test suite, which is what the changes in the `cassettes` directory are.

Murphydbuffalo committed:
Don't even attempt API request if required data is missing

Murphydbuffalo committed:
Make missing data logging consistent with other logging

Murphydbuffalo committed:
In log message specify which piece of data is missing

Murphydbuffalo committed:
DRY out `try_conversion` and `try_assignment`

Murphydbuffalo committed:
Include API response error message in logging

Murphydbuffalo committed:
Bump package version

Murphydbuffalo committed:
Update README

Murphydbuffalo committed:
Specify content type

Murphydbuffalo committed:
Fix linter violations

Murphydbuffalo committed:
Re-record test requests with added Content-Type header

wfm commented on src/hyp_client/v1.py:
```diff
@@ -8,44 +8,55 @@ def __init__(self, access_token, session=requests.Session()):
+
+        if response["message"] == "success":
+            self.logger.info(f'{method_name} successful for participant {participant_id} in experiment {experiment_id}.')
+            if method_name == "assignment":
```
This `if` statement isn't extensible. If you were to add another API method with a `try_` method, you'd have to add two more lines of code to this `if` statement to handle the new method.

You could pass the field name as a parameter and use that to extract the data from the response. That way you eliminate the use of the `method_name`.

e.g.:
`def try_api_call(self, method_name, participant_id, experiment_id, response_field, fallback):`
...
`return response["payload"][field_name]`

Murphydbuffalo replied:
I generally tend to not optimize for extensibility right away. If we continue to add conditionals to this code it'll be signal that there is a more general abstraction that should be leaned into. Right now it's not clear that there will be more API calls to support and what the mechanics of those API calls will be. If that does turn out to be the case then I'll be more inclined to make this more dynamic.

There is certainly a lot of subjectivity here, I'm not saying that it would definitely be wrong to handle this as you suggest, just that I don't think the current imperative code is problematic.

Eg it's not as if this conditional is very large or complex, resulting in difficult to understand code.

wfm reviewed (commented):
I left a comment in the code, at line 50 of v1.py. Also, I posted a comment in Discord about the __init__ function. The linter I am using flagged the argument default for the session parameter:
`def __init__(self, access_token, session=requests.Session()):`
It said, 

> Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.

Murphydbuffalo committed:
Don't use functions calls as default argument values

The function you're using to provide the default argument value
will only be called once, at the time the code is evaluated.

Murphydbuffalo commented:
Thanks for the feedback! When reviewing a PR the norm is not to close it if you have feedback, but to let the person who made the PR respond to and potentially address that feedback :).

Murphydbuffalo commented:
> I left a comment in the code, at line 50 of v1.py. Also, I posted a comment in Discord about the **init** function. The linter I am using flagged the argument default for the session parameter: `def __init__(self, access_token, session=requests.Session()):` It said,
> 
> > Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.

Good call, I've updated this 👍

Murphydbuffalo commented:
I need to bump the package version number and re-deploy it to pypl with the changes from [614a73b](https://github.com/Murphydbuffalo/hyp_python_client/pull/8/commits/614a73b52c74ae295a9637fa187943c9137af31f)
